File: problems/tier2-intermediate/021-ride-sharing/solution.java
// Ride Sharing — Solution (Java)
import java.util.*;

class RideOp {
    public String kind;
    public String s1, s2, s3, s4;
    public int i1, i2;
    public RideOp(String kind, String s1, String s2, String s3, String s4, int i1, int i2) {
        this.kind = kind; this.s1 = s1; this.s2 = s2; this.s3 = s3; this.s4 = s4;
        this.i1 = i1; this.i2 = i2;
    }
}

class User {
    public String id, name;
    public int ridesOffered, ridesTaken;
    public User(String id, String name) { this.id = id; this.name = name; }
}

class Vehicle {
    public String id, ownerId, model, regNumber;
    public Vehicle(String id, String ownerId, String model, String reg) {
        this.id = id; this.ownerId = ownerId; this.model = model; this.regNumber = reg;
    }
}

class Ride {
    public String id;
    public String driverId;
    public String vehicleId;
    public String origin;
    public String destination;
    public int totalSeats;
    public int availableSeats;
    public boolean active;
    public Ride(String id, String driverId, String vehicleId, String origin, String dest, int seats) {
        this.id = id; this.driverId = driverId; this.vehicleId = vehicleId;
        this.origin = origin; this.destination = dest;
        this.totalSeats = seats; this.availableSeats = seats; this.active = true;
    }
}

interface RideSelectionStrategy {
    Ride select(List<Ride> candidates, int seatsNeeded, String preference);
}

class MostVacantStrategy implements RideSelectionStrategy {
    @Override public Ride select(List<Ride> candidates, int seatsNeeded, String preference) {
        Ride best = null;
        for (Ride r : candidates) {
            if (r.availableSeats >= seatsNeeded) {
                if (best == null || r.availableSeats > best.availableSeats) best = r;
            }
        }
        return best;
    }
}

class PreferredVehicleStrategy implements RideSelectionStrategy {
    private final Map<String, Vehicle> vehicleStore;
    public PreferredVehicleStrategy(Map<String, Vehicle> vs) { this.vehicleStore = vs; }
    @Override public Ride select(List<Ride> candidates, int seatsNeeded, String preference) {
        for (Ride r : candidates) {
            if (r.availableSeats >= seatsNeeded) {
                Vehicle v = vehicleStore.get(r.vehicleId);
                if (v != null && v.model.equals(preference)) return r;
            }
        }
        return null;
    }
}

class RideService {
    Map<String, User> users = new LinkedHashMap<>();
    Map<String, Vehicle> vehicles = new LinkedHashMap<>();
    Map<String, Ride> rides = new LinkedHashMap<>();
    Map<String, String> activeVehicles = new LinkedHashMap<>();
    int rideCounter = 0;

    public void addUser(String name) {
        if (users.containsKey(name)) return;
        users.put(name, new User(name, name));
    }

    public void addVehicle(String userName, String model, String regNumber) {
        if (!users.containsKey(userName)) return;
        vehicles.put(regNumber, new Vehicle(regNumber, userName, model, regNumber));
    }

    public String offerRide(String userName, String origin, String dest, int seats, String vehicleRegNumber) {
        if (!users.containsKey(userName)) return "";
        Vehicle v = vehicles.get(vehicleRegNumber);
        if (v == null) return "";
        if (!v.ownerId.equals(userName)) return "";
        if (activeVehicles.containsKey(vehicleRegNumber)) return "";

        String rideId = "RIDE-" + (++rideCounter);
        rides.put(rideId, new Ride(rideId, userName, vehicleRegNumber, origin, dest, seats));
        activeVehicles.put(vehicleRegNumber, rideId);
        users.get(userName).ridesOffered++;
        return rideId;
    }

    public String selectRide(String passengerName, String origin, String dest, int seats,
                              RideSelectionStrategy strategy, String preference) {
        if (!users.containsKey(passengerName)) return "";

        List<Ride> candidates = new ArrayList<>();
        for (Ride r : rides.values()) {
            if (r.active && r.origin.equals(origin) && r.destination.equals(dest)
                    && r.availableSeats >= seats && !r.driverId.equals(passengerName)) {
                candidates.add(r);
            }
        }
        Ride selected = strategy.select(candidates, seats, preference);
        if (selected != null) {
            selected.availableSeats -= seats;
            users.get(passengerName).ridesTaken++;
            return selected.id;
        }
        return "";
    }

    public void endRide(String rideId) {
        Ride r = rides.get(rideId);
        if (r == null || !r.active) return;
        r.active = false;
        activeVehicles.remove(r.vehicleId);
    }

    public Map<String, Vehicle> getVehicles() { return vehicles; }

    public boolean hasUser(String name) { return users.containsKey(name); }
    public boolean hasVehicle(String reg) { return vehicles.containsKey(reg); }
    public boolean hasRide(String id) { return rides.containsKey(id); }
    public User getUser(String name) { return users.get(name); }
    public Ride getRide(String id) { return rides.get(id); }
}

public class Solution {
    public static List<String> ride_simulate(List<RideOp> ops) {
        List<String> out = new ArrayList<>();
        RideService svc = new RideService();
        String[] rideSlots = new String[32];
        for (int i = 0; i < 32; i++) rideSlots[i] = "";

        for (RideOp op : ops) {
            String k = op.kind;
            if ("new".equals(k)) {
                svc = new RideService();
                for (int i = 0; i < 32; i++) rideSlots[i] = "";
                out.add("ok");
            } else if ("add_user".equals(k)) {
                svc.addUser(op.s1); out.add("ok");
            } else if ("add_veh".equals(k)) {
                svc.addVehicle(op.s1, op.s2, op.s3); out.add("ok");
            } else if ("offer".equals(k)) {
                String rid = svc.offerRide(op.s1, op.s2, op.s3, op.i1, op.s4);
                if (op.i2 >= 0 && op.i2 < rideSlots.length) rideSlots[op.i2] = rid;
                out.add(rid);
            } else if ("ride_active".equals(k)) {
                String rid = rideSlots[op.i2];
                out.add(svc.hasRide(rid) && svc.getRide(rid).active ? "yes" : "no");
            } else if ("ride_origin".equals(k)) {
                out.add(svc.hasRide(rideSlots[op.i2]) ? svc.getRide(rideSlots[op.i2]).origin : "");
            } else if ("ride_dest".equals(k)) {
                out.add(svc.hasRide(rideSlots[op.i2]) ? svc.getRide(rideSlots[op.i2]).destination : "");
            } else if ("ride_total".equals(k)) {
                out.add(Integer.toString(svc.hasRide(rideSlots[op.i2]) ? svc.getRide(rideSlots[op.i2]).totalSeats : -1));
            } else if ("ride_avail".equals(k)) {
                out.add(Integer.toString(svc.hasRide(rideSlots[op.i2]) ? svc.getRide(rideSlots[op.i2]).availableSeats : -1));
            } else if ("ride_driver".equals(k)) {
                out.add(svc.hasRide(rideSlots[op.i2]) ? svc.getRide(rideSlots[op.i2]).driverId : "");
            } else if ("select_mv".equals(k)) {
                String rid = svc.selectRide(op.s1, op.s2, op.s3, op.i1, new MostVacantStrategy(), "");
                if (op.i2 >= 0 && op.i2 < rideSlots.length) rideSlots[op.i2] = rid;
                out.add(rid);
            } else if ("select_pv".equals(k)) {
                String rid = svc.selectRide(op.s1, op.s2, op.s3, op.i1,
                        new PreferredVehicleStrategy(svc.getVehicles()), op.s4);
                if (op.i2 >= 0 && op.i2 < rideSlots.length) rideSlots[op.i2] = rid;
                out.add(rid);
            } else if ("end".equals(k)) {
                svc.endRide(rideSlots[op.i2]); out.add("ok");
            } else if ("end_id".equals(k)) {
                svc.endRide(op.s1); out.add("ok");
            } else if ("user_offered".equals(k)) {
                out.add(svc.hasUser(op.s1) ? Integer.toString(svc.getUser(op.s1).ridesOffered) : "0");
            } else if ("user_taken".equals(k)) {
                out.add(svc.hasUser(op.s1) ? Integer.toString(svc.getUser(op.s1).ridesTaken) : "0");
            } else if ("has_user".equals(k)) {
                out.add(svc.hasUser(op.s1) ? "yes" : "no");
            } else if ("has_vehicle".equals(k)) {
                out.add(svc.hasVehicle(op.s1) ? "yes" : "no");
            } else if ("has_ride".equals(k)) {
                out.add(svc.hasRide(rideSlots[op.i2]) ? "yes" : "no");
            } else {
                out.add("unknown:" + k);
            }
        }
        return out;
    }
}

File: problems/tier2-intermediate/021-ride-sharing/solution.js
"use strict";

class RideOp {
  constructor(kind, s1 = "", s2 = "", s3 = "", s4 = "", i1 = 0, i2 = 0) {
    this.kind = kind;
    this.s1 = s1;
    this.s2 = s2;
    this.s3 = s3;
    this.s4 = s4;
    this.i1 = i1;
    this.i2 = i2;
  }
}

class User {
  constructor(id, name) {
    this.id = id;
    this.name = name;
    this.rides_offered = 0;
    this.rides_taken = 0;
  }
}

class Vehicle {
  constructor(id, owner_id, model, reg_number) {
    this.id = id;
    this.owner_id = owner_id;
    this.model = model;
    this.reg_number = reg_number;
  }
}

class Ride {
  constructor(id, driver_id, vehicle_id, origin, destination, total_seats) {
    this.id = id;
    this.driver_id = driver_id;
    this.vehicle_id = vehicle_id;
    this.origin = origin;
    this.destination = destination;
    this.total_seats = total_seats;
    this.available_seats = total_seats;
    this.active = true;
  }
}

class MostVacantStrategy {
  select(candidates, seats_needed, preference) {
    let best = null;
    for (const ride of candidates) {
      if (ride.available_seats >= seats_needed) {
        if (best === null || ride.available_seats > best.available_seats) {
          best = ride;
        }
      }
    }
    return best;
  }
}

class PreferredVehicleStrategy {
  constructor(vehicle_store) {
    this.vehicle_store = vehicle_store;
  }
  select(candidates, seats_needed, preference) {
    for (const ride of candidates) {
      if (ride.available_seats >= seats_needed) {
        const v = this.vehicle_store.has(ride.vehicle_id)
          ? this.vehicle_store.get(ride.vehicle_id)
          : undefined;
        if (v !== undefined && v.model === preference) {
          return ride;
        }
      }
    }
    return null;
  }
}

class RideService {
  constructor() {
    this.users = new Map();
    this.vehicles = new Map();
    this.rides = new Map();
    this.active_vehicles = new Map(); // regNumber -> rideId
    this.ride_counter = 0;
  }

  add_user(name) {
    if (this.users.has(name)) return;
    this.users.set(name, new User(name, name));
  }

  add_vehicle(user_name, model, reg_number) {
    if (!this.users.has(user_name)) return;
    this.vehicles.set(reg_number, new Vehicle(reg_number, user_name, model, reg_number));
  }

  offer_ride(user_name, origin, dest, seats, vehicle_reg_number) {
    if (!this.users.has(user_name)) return "";
    if (!this.vehicles.has(vehicle_reg_number)) return "";
    if (this.vehicles.get(vehicle_reg_number).owner_id !== user_name) return "";
    if (this.active_vehicles.has(vehicle_reg_number)) return "";
    this.ride_counter += 1;
    const ride_id = "RIDE-" + String(this.ride_counter);
    this.rides.set(ride_id, new Ride(ride_id, user_name, vehicle_reg_number, origin, dest, seats));
    this.active_vehicles.set(vehicle_reg_number, ride_id);
    this.users.get(user_name).rides_offered += 1;
    return ride_id;
  }

  select_ride(passenger_name, origin, dest, seats, strategy, preference = "") {
    if (!this.users.has(passenger_name)) return "";
    const candidates = [];
    for (const ride of this.rides.values()) {
      if (ride.active && ride.origin === origin && ride.destination === dest
          && ride.available_seats >= seats && ride.driver_id !== passenger_name) {
        candidates.push(ride);
      }
    }
    const selected = strategy.select(candidates, seats, preference);
    if (selected !== null && selected !== undefined) {
      selected.available_seats -= seats;
      this.users.get(passenger_name).rides_taken += 1;
      return selected.id;
    }
    return "";
  }

  end_ride(ride_id) {
    const ride = this.rides.has(ride_id) ? this.rides.get(ride_id) : null;
    if (ride === null) return;
    if (!ride.active) return;
    ride.active = false;
    if (this.active_vehicles.has(ride.vehicle_id)) {
      this.active_vehicles.delete(ride.vehicle_id);
    }
  }

  has_user(name) { return this.users.has(name); }
  has_vehicle(reg) { return this.vehicles.has(reg); }
  has_ride(rid) { return this.rides.has(rid); }
  get_user(name) { return this.users.get(name); }
  get_ride(rid) { return this.rides.get(rid); }
  get_vehicles() { return this.vehicles; }
}

function ride_simulate(ops) {
  const out = [];
  let svc = new RideService();
  let ride_slots = new Array(32).fill("");
  for (const op of ops) {
    const k = op.kind;
    if (k === "new") {
      svc = new RideService();
      ride_slots = new Array(32).fill("");
      out.push("ok");
    } else if (k === "add_user") {
      svc.add_user(op.s1);
      out.push("ok");
    } else if (k === "add_veh") {
      svc.add_vehicle(op.s1, op.s2, op.s3);
      out.push("ok");
    } else if (k === "offer") {
      const rid = svc.offer_ride(op.s1, op.s2, op.s3, op.i1, op.s4);
      if (0 <= op.i2 && op.i2 < ride_slots.length) ride_slots[op.i2] = rid;
      out.push(rid);
    } else if (k === "ride_active") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) && svc.get_ride(rid).active ? "yes" : "no");
    } else if (k === "ride_origin") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) ? svc.get_ride(rid).origin : "");
    } else if (k === "ride_dest") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) ? svc.get_ride(rid).destination : "");
    } else if (k === "ride_total") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) ? String(svc.get_ride(rid).total_seats) : "-1");
    } else if (k === "ride_avail") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) ? String(svc.get_ride(rid).available_seats) : "-1");
    } else if (k === "ride_driver") {
      const rid = ride_slots[op.i2];
      out.push(svc.has_ride(rid) ? svc.get_ride(rid).driver_id : "");
    } else if (k === "select_mv") {
      const rid = svc.select_ride(op.s1, op.s2, op.s3, op.i1, new MostVacantStrategy());
      if (0 <= op.i2 && op.i2 < ride_slots.length) ride_slots[op.i2] = rid;
      out.push(rid);
    } else if (k === "select_pv") {
      const rid = svc.select_ride(op.s1, op.s2, op.s3, op.i1,
        new PreferredVehicleStrategy(svc.get_vehicles()), op.s4);
      if (0 <= op.i2 && op.i2 < ride_slots.length) ride_slots[op.i2] = rid;
      out.push(rid);
    } else if (k === "end") {
      svc.end_ride(ride_slots[op.i2]);
      out.push("ok");
    } else if (k === "end_id") {
      svc.end_ride(op.s1);
      out.push("ok");
    } else if (k === "user_offered") {
      out.push(svc.has_user(op.s1) ? String(svc.get_user(op.s1).rides_offered) : "0");
    } else if (k === "user_taken") {
      out.push(svc.has_user(op.s1) ? String(svc.get_user(op.s1).rides_taken) : "0");
    } else if (k === "has_user") {
      out.push(svc.has_user(op.s1) ? "yes" : "no");
    } else if (k === "has_vehicle") {
      out.push(svc.has_vehicle(op.s1) ? "yes" : "no");
    } else if (k === "has_ride") {
      out.push(svc.has_ride(ride_slots[op.i2]) ? "yes" : "no");
    } else {
      out.push("unknown:" + k);
    }
  }
  return out;
}

module.exports = { RideOp, ride_simulate };
